libraries/resource_providers/api500/c7000/volume_provider.rb

```diff
@@ -0,0 +1,103 @@
+          else
+            validate_required_properties(:storage_system, :storage_pool)
+            storage_system_uri = load_storage_system
+            @item['storagePoolUri'] = load_resource(:StoragePool, { name: @new_resource.storage_pool, storageSystemUri: storage_system_uri }, 'uri')
+            @item['snapshotPoolUri'] = load_resource(:StoragePool, { name: @new_resource.snapshot_pool, storageSystemUri: storage_system_uri }, 'uri') if @new_resource.snapshot_pool
+          end
+          set_properties unless @item.exists?
+        end
+
+        def load_storage_system
+          data = {
+            hostname: @new_resource.storage_system,
+            name: @new_resource.storage_system
+          }
+          load_resource(:StorageSystem, data, 'uri')
+        end
+
+        def load_volume_template
+          template = resource_named(:VolumeTemplate).new(@item.client, name: @new_resource.volume_template)
+          @item.set_storage_volume_template(template)
+          @item['storagePoolUri'] = template['storagePoolUri']
+          @item['snapshotPoolUri'] = template['properties']['snapshotPool']['default'] if template['family'] == 'StoreServ'
+        end
+
+        def create_from_snapshot
+          validate_required_properties(:properties)
+          raise("#{@resource_name} '#{@name}' not found!") unless @item.retrieve!
+          properties = convert_keys(@new_resource.properties, :to_s)
+          properties['isShareable'] ||= @new_resource.is_shareable
+          return Chef::Log.info "Volume '#{properties['name']}' already exists" if resource_named(:Volume).new(@item.client, name: properties['name']).retrieve!
+          snapshot_name = properties['snapshotName']
+          properties.delete('snapshotName')
+          volume_template = nil
+          volume_template = resource_named(:VolumeTemplate).new(@item.client, name: @new_resource.volume_template) if @new_resource.volume_template
+          Chef::Log.info "Creating #{@resource_name} '#{properties['name']}' from snapshot #{snapshot_name}"
+          @context.converge_by "Created #{@resource_name} '#{properties['name']}' from snapshot #{snapshot_name}" do
+            @item.create_from_snapshot(snapshot_name, properties, volume_template, @new_resource.is_permanent)
+          end
+        end
+
+        def add_or_edit
+          validate_required_properties(:storage_system, :device_volume_name, :is_shareable)
+          options = Marshal.load(Marshal.dump(@item.data))
+          return Chef::Log.info "#{@resource_name} '#{@name}' already exists" if @item.exists?
+          storage_system = resource_named(:StorageSystem).new(@item.client, name: @new_resource.storage_system)
+          Chef::Log.info "Adding #{@resource_name} '#{@name}'"
+          @context.converge_by "Added #{@resource_name} '#{@name}'" do
+            resource_named(:Volume).add(@item.client, storage_system, @new_resource.device_volume_name, @new_resource.is_shareable, options)
```

Comment:
Gotta say I don't particularly like this `add` statement, since it differs so much from the usual `add(client, data)`. 

I understand this is implemented this way in the SDK at the moment, but it would probably be better for standardization purposes to keep it in the usual format, and extract the info from data. Or have helpers to help build the final object. 

It just feels off to me, out of the expected for the `add` action to expect these many info and not get a good body for a post directly.